### PR TITLE
feat: implement user collaboration features

### LINKED ISSUE #719: feature: Collaborate workflow

### LINKED ISSUE #1003: feat: Integrate the API into the collaboration

### PR DESCRIPTION
**Pull request by OgDev-01**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description
This PR:

- Adds users collaboration feature
- Fixes contributor profile header layout break
- Forces a user to signup when they click on collaborate button and redirect back to where they left off
- implement login to show and hide the collaboration tab based on user's preferences 
- Requesting collaborations is currently available to paid users only.

## Related Tickets & Documents
Closes #719
Closes #1003 

## Mobile & Desktop Screenshots/Recordings
image

image

image

image

image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

**Comment by bdougie** [comment]
@brandonroberts any chance we can implement feature flagging for this feature? I really think the UI is ready for prime time, but it would be nice to show the feature behind the scenes. 

Perhaps we can make this for paid users only for now.

**Comment by brandonroberts** [comment]
@bdougie yes, we can use the same logic for determining access to reports to restrict seeing this. We can land this and follow-up with the role check.

File: components/molecules/ContributorProfileHeader/contributor-profile-header.tsx [hunk 1]
```diff
@@ -1,13 +1,25 @@
-import React, { useState, useEffect } from "react";
```
Comment (brandonroberts) [inline components/molecules/ContributorProfileHeader/contributor-profile-header.tsx:1]: Organize the imports while we're in here

File: components/molecules/ContributorProfileHeader/contributor-profile-header.tsx [hunk 2]
```diff
@@ -43,6 +68,29 @@ const ContributorProfileHeader = ({
...
+    if (message && username) {
+      setLoading(true);
+      await requestCollaboration({ username, message });
+      setIsDialogOpen(false);
+      setLoading(false);
+      setMessage("");
+    }
+  };
+
+  const handleTextAreaInputChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
```
Comment (brandonroberts) [inline components/molecules/ContributorProfileHeader/contributor-profile-header.tsx:83]: We've done in this in a couple of places now. We should make this a reusable component.

File: components/organisms/CollaborationRequestWrapper/collaboration-requests-wrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    return "Yesterday";
+  } else {
+    return formatDistanceToNowStrict(date, { addSuffix: true });
+  }
+}
+
+const CollaborationRequestsWrapper = () => {
+  const { data, updateCollaborationStatus, deleteCollaborationRequest } = useUserCollaborations();
+  let currentDate: string;
+  console.log(data);
```
Comment (brandonroberts) [inline components/organisms/CollaborationRequestWrapper/collaboration-requests-wrapper.tsx:20]:
```suggestion

```

File: components/organisms/CollaborationRequestWrapper/collaboration-requests-wrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        </p>
+      );
+    }
+    return;
+  };
+
+  return (
+    <div>
+      {data && data.length > 0 ? (
+        data
```
Comment (brandonroberts) [inline components/organisms/CollaborationRequestWrapper/collaboration-requests-wrapper.tsx:39]: Create a new array if you're going to sort in place
```suggestion
        data.slice()
```

File: lib/hooks/useSupabaseAuth.ts [hunk 5]
```diff
@@ -6,10 +6,10 @@ import { supabase } from "../utils/supabase";
 
 const useSupabaseAuth = (loadSession = false) => {
   const store = useStore();
-  const user = useStore(state => state.user);
-  const sessionToken = useStore(state => state.sessionToken);
-  const providerToken = useStore(state => state.providerToken);
-  const userId = useStore(state => state.userId);
+  const user = useStore((state) => state.user);
```
Comment (brandonroberts) [inline lib/hooks/useSupabaseAuth.ts:9]: Did you make any changes to this file other than formatting?

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: lib/hooks/useSupabaseAuth.ts [hunk 6]
```diff
@@ -37,17 +39,18 @@ const useSupabaseAuth = (loadSession = false) => {
...
-    signIn: (data: SignInWithOAuthCredentials) => supabase.auth.signInWithOAuth({
-      ...data,
-      options: {
-        redirectTo: process.env.NEXT_PUBLIC_BASE_URL ?? "/"
-      }
-    }),
+    signIn: (data: SignInWithOAuthCredentials) =>
+      supabase.auth.signInWithOAuth({
+        ...data,
+        options: data.options ?? {
```
Comment (OgDev-01) [inline lib/hooks/useSupabaseAuth.ts:45]: Yeah I added users passed in option here

**Comment by bdougie** [comment]
Reviewing this and not seeing the dropdown anymore. 

Screen Shot 2023-05-31 at 12 26 56 PM

**Comment by OgDev-01** [comment]
> Reviewing this and not seeing the dropdown anymore.
> 
> Screen Shot 2023-05-31 at 12 26 56 PM

Should the dropdown also be used in the desktop view? I was thinking we use it on mobile alone.

**Comment by bdougie** [comment]
We need it on desktop. 3 buttons is a bit much. We will ultimately need a better design eventually, but the dropdown works.

**Review by bdougie — CHANGES_REQUESTED** [review]
Per slack, we need to add a "paid user" feature flag for the collaboration button. This will allow us to test the feature and iterate.

**Review by brandonroberts — APPROVED** [review]
LGTM

**Pull request merged**